include/wil/registry_helpers.h

```diff
@@ -0,0 +1,947 @@
+                {
+                    return const_cast<wchar_t*>(value.data());
+                }
+#endif // #if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+
+#if defined(__WIL_OLEAUTO_H_)
+                inline void* get_buffer(const BSTR& value) WI_NOEXCEPT
+                {
+                    return value;
+                }
+                inline void* get_buffer(const ::wil::unique_bstr& value) WI_NOEXCEPT
+                {
+                    return value.get();
+                }
+#endif // #if defined(__WIL_OLEAUTO_H_)
+
+#if defined(__WIL_OLEAUTO_H_STL)
+                inline void* get_buffer(const ::wil::shared_bstr& value) WI_NOEXCEPT
+                {
+                    return value.get();
+                }
+#endif // #if defined(__WIL_OLEAUTO_H_STL)
+
+#if defined(__WIL_OBJBASE_H_)
+                inline void* get_buffer(const ::wil::unique_cotaskmem_string& value) WI_NOEXCEPT
+                {
+                    return value.get();
+                }
+#endif // defined(__WIL_OBJBASE_H_)
+
+#if defined(__WIL_OBJBASE_H_STL)
+                inline void* get_buffer(const ::wil::shared_cotaskmem_string& value) WI_NOEXCEPT
+                {
+                    return value.get();
+                }
+#endif // #if defined(__WIL_OBJBASE_H_STL)
+
+#if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+                inline DWORD get_buffer_size(const ::std::vector<BYTE>& value) WI_NOEXCEPT
+                {
+                    return static_cast<DWORD>(value.size());
+                }
+                inline DWORD get_buffer_size(const ::std::vector<wchar_t>& value) WI_NOEXCEPT
+                {
+                    return static_cast<DWORD>(value.size()) * sizeof(wchar_t);
+                }
+#endif // #if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+
+#if defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+                inline DWORD get_buffer_size(const ::std::wstring& string) WI_NOEXCEPT
+                {
+                    return static_cast<DWORD>(string.size() * sizeof(wchar_t));
+                }
+#endif // #if defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+
+                constexpr DWORD get_buffer_size(int32_t) WI_NOEXCEPT
+                {
+                    return static_cast<DWORD>(sizeof(int32_t));
+                }
+
+                constexpr DWORD get_buffer_size(uint32_t) WI_NOEXCEPT
+                {
+                    return static_cast<DWORD>(sizeof(uint32_t));
+                }
+
+                constexpr DWORD get_buffer_size(long) WI_NOEXCEPT
+                {
+                    return static_cast<DWORD>(sizeof(long));
+                }
+
+                constexpr DWORD get_buffer_size(unsigned long) WI_NOEXCEPT
+                {
+                    return static_cast<DWORD>(sizeof(unsigned long));
+                }
+
+                constexpr DWORD get_buffer_size(int64_t) WI_NOEXCEPT

```

Comment:
Consider converting these overloads into a single templatized thing with std::is_scalar_v<T> in one branch and "is like a string" in the other.

Reply:
I played with this a while, but couldn't find exactly what met requirements :(